### PR TITLE
Support visibility dual write to different ES indices

### PR DESCRIPTION
**Pull request by meiliang86**
**What changed?**
Support visibility dual write to different ES indices
Add additional dynamic configs to control visibility dual write and read behavior

**Why?**
To support seamless switch to new ES visibility index

**How did you test it?**
Local test with 2 ES indices

**Potential risks**
No. The old index should only be deleted after the new index is verified working.

**Is hotfix candidate?**
No

### CODE REVIEW

File: common/persistence/visibility/factory.go [hunk 1]
```diff
@@ -138,7 +179,7 @@ func NewStandardManager(
 }
 
 func NewAdvancedManager(
-	defaultIndexName string,
+	indexName string,
```
Comment (alexshtin) [inline common/persistence/visibility/factory.go:182]: The reason is was called `defaultIndexName` because there is partial support for index per namespace. Metadata has `indexName` field which is not propagated and used yet. The one from config supposed to be default index name if it is not configured on namespace level. I think you should rename it back.

File: common/dynamicconfig/constants.go [hunk 2]
```diff
@@ -70,7 +70,9 @@ var Keys = map[Key]string{
 	AdvancedVisibilityPersistenceMaxReadQPS:  "system.advancedVisibilityPersistenceMaxReadQPS",
 	AdvancedVisibilityPersistenceMaxWriteQPS: "system.advancedVisibilityPersistenceMaxWriteQPS",
 	AdvancedVisibilityWritingMode:            "system.advancedVisibilityWritingMode",
+	EnableWriteToSecondaryVisibility:         "system.enableWriteToSecondaryVisibility",
```
Comment (alexshtin) [inline common/dynamicconfig/constants.go:73]: may be:
```suggestion
	EnableWriteToSecondaryAdvancedVisibility:         "system.enableWriteToSecondaryAdvancedVisibility",
```

File: common/persistence/visibility/factory.go [hunk 3]
```diff
@@ -44,7 +44,7 @@ func NewManager(
 	persistenceCfg config.Persistence,
 	persistenceResolver resolver.ServiceResolver,
 
-	defaultIndexName string,
+	esConfig *esclient.Config,
```
Comment (alexshtin) [inline common/persistence/visibility/factory.go:47]: I would pass both index names here rather than taking dependency on the config.

File: common/persistence/visibility/store/elasticsearch/client/config.go [hunk 4]
```diff
@@ -80,12 +81,18 @@ type (
 // GetVisibilityIndex return visibility index name from Elasticsearch config or empty string if it is not defined.
 func (cfg *Config) GetVisibilityIndex() string {
 	if cfg == nil {
-		// Empty string is be used as default index name when Elasticsearch is not configured.
```
Comment (alexshtin) [inline common/persistence/visibility/store/elasticsearch/client/config.go:83]: It was important comment. Why did you remove it? It had grammar errors though.

File: common/persistence/visibility/visibility_manager_dual.go [hunk 5]
```diff
@@ -28,38 +28,83 @@ import (
...
+	}
+
+	sqlToESManagerSelector struct {
 		enableAdvancedVisibilityRead  dynamicconfig.BoolPropertyFnWithNamespaceFilter
 		advancedVisibilityWritingMode dynamicconfig.StringPropertyFn
+		stdVisibilityManager          manager.VisibilityManager
+		advVisibilityManager          manager.VisibilityManager
+	}
+
+	esManagerSelector struct {
```
Comment (alexshtin) [inline common/persistence/visibility/visibility_manager_dual.go:55]: Please place all these structs and their methods in separate files. I hate to mix structs with methods in one file (I know we have many of those but it is just a legacy and I am constantly fixing it).

File: common/persistence/visibility/store/elasticsearch/visibility_store.go [hunk 6]
```diff
@@ -598,14 +600,22 @@ func (s *visibilityStore) buildSearchParameters(
...
+			params.Sorter = append(params.Sorter, elastic.NewFieldSort(searchattribute.StartTime).Desc())
+		} else {
+			params.Sorter = append(params.Sorter, elastic.NewFieldSort(searchattribute.CloseTime).Desc())
+		}
 
-	// RunID is explicit tiebreaker.
-	params.Sorter = append(params.Sorter, elastic.NewFieldSort(searchattribute.RunID).Desc())
+		// RunID is explicit tiebreaker.
+		params.Sorter = append(params.Sorter, elastic.NewFieldSort(searchattribute.RunID).Desc())
+	}
```
[the remark this answers is not in the document]
Comment (meiliang86) [reply common/persistence/visibility/store/elasticsearch/visibility_store.go:618]: Changed to use the default sorting order in all cases.

File: common/persistence/visibility/factory.go (hunk 1 above)
Comment (meiliang86) [reply common/persistence/visibility/factory.go:182]: reverted

File: common/dynamicconfig/constants.go (hunk 2 above)
Comment (meiliang86) [reply common/dynamicconfig/constants.go:73]: done

File: common/persistence/visibility/store/elasticsearch/client/config.go (hunk 4 above)
Comment (meiliang86) [reply common/persistence/visibility/store/elasticsearch/client/config.go:83]: I will add this back. But how is it supposed to work if user does not specify index name?

File: common/dynamicconfig/constants.go [hunk 7]
```diff
@@ -383,8 +385,12 @@ const (
 	AdvancedVisibilityPersistenceMaxWriteQPS
 	// AdvancedVisibilityWritingMode is key for how to write to advanced visibility
 	AdvancedVisibilityWritingMode
+	// EnableWriteToSecondaryAdvancedVisibility is the config to enable write to secondary visibility for elastic search
```
Comment (alexshtin) [inline common/dynamicconfig/constants.go:388]:
```suggestion
	// EnableWriteToSecondaryAdvancedVisibility is the config to enable write to secondary visibility for Elasticsearch.
```

File: common/dynamicconfig/constants.go [hunk 8]
```diff
@@ -383,8 +385,12 @@ const (
...
+	EnableWriteToSecondaryAdvancedVisibility
 	// EnableReadVisibilityFromES is key for enable read from elastic search
 	EnableReadVisibilityFromES
+	// EnableReadFromSecondaryAdvancedVisibility is the config to enable read from secondary elastic search
```
Comment (alexshtin) [inline common/dynamicconfig/constants.go:392]:
```suggestion
	// EnableReadFromSecondaryAdvancedVisibility is the config to enable read from secondary Elasticsearch.
```

**Review by alexshtin — APPROVED** [review]
